### PR TITLE
RFC: Range Repair

### PR DESCRIPTION
**Pull request by BramGruneir**
This RFC is designed to work in conjunction with #2153

### CODE REVIEW

File: docs/RFCS/range_repair.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
+- Feature Name: Range Repair
```
Comment (vivekmenezes) [inline docs/RFCS/range_repair.md]: Hmmm, it's not really repairing a range. Perhaps we should call this "Range replica reincarnation"

Comment (BramGruneir) [inline docs/RFCS/range_repair.md]: Yeah, it's just pruning dead replicas and suggesting adding new ones.  I could call it the `Replica Pruning Service`.

File: docs/RFCS/range_repair.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+- Start Date: 2015-08-19
+- RFC PR:
+- Cockroach Issue: #2149
+
+# Summary
+
+Add a new `Repair` service on each node that monitors all other nodes via
+gossip and determines if any of them are dead. When a node dies, the service
+looks through all the local ranges to see they have any replicas on the dead
+node. If any exist, remove the dead replica and add the local replica to the
```
Comment (petermattis) [inline docs/RFCS/range_repair.md]: In practice, I think you'll want to queue up this work. Imagine a cluster with 10 nodes where one node dies temporarily. This would queue up 1/10th of the ranges for repair. Then the dead node comes back (e.g. a sysadmin kicked it or a machine reboot finished). You don't want to have to run through all of the affect ranges again.

File: docs/RFCS/range_repair.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+1. mark the node as dead
+2. check all the ranges with local replicas to see if any had a replica on this
+   now dead node
+3. for any ranges that do,
+   1. remove replica from the dead node
+     - if this fails, add a minute long delay to the `timeUntilNextDead` for
+       the node and skip to step 4
+   2. add the replica to the replication queue
+4. calculate the next `timeUntilNextDead` to wake up the service and add the
```
Comment (vivekmenezes) [inline docs/RFCS/range_repair.md]: this step is always done. You need not need to mention it here.

File: docs/RFCS/range_repair.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+Repair will maintain a timespan `timeUntilNextDead` which is calculated by
+taking the nearest last update time for all the nodes and adding
+`timeoutUntilNodeDead` and the node ID associated with the timeout.
+
+Repair will trigger on `timeUntilNextDead` which when triggered checks to see
+if that node has not been updated.
+
+If the node is dead:
+
+1. mark the node as dead
```
Comment (mrtracy) [inline docs/RFCS/range_repair.md]: Where is it being marked as dead?  Is this information relevant to some other service?

File: docs/RFCS/range_repair.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+considered dead. The default value for this will be 5 minutes.
+
+## Repair
+Add a new service called `Repair` that starts when the node is started.
+This new service will run until the stopper is called and have access to both
+gossip and the list of stores (in order to get the list of replicas on those
+stores).
+
+Repair will also maintain a map of node IDs to last update time. To maintain
+this map, a callback from gossip for node descriptors will be added.
```
Comment (mrtracy) [inline docs/RFCS/range_repair.md]: Is there an existing gossiped value of "lastUpdateTime"?  If not, we need to have each node place such a value into gossip on some interval.

File: docs/RFCS/range_repair.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+`timeoutUntilNodeDead` and the node ID associated with the timeout.
+
+Repair will trigger on `timeUntilNextDead` which when triggered checks to see
+if that node has not been updated.
+
+If the node is dead:
+
+1. mark the node as dead
+2. check all the ranges with local replicas to see if any had a replica on this
+   now dead node
```
Comment (mrtracy) [inline docs/RFCS/range_repair.md]: This sounds like a possibly intensive process; should we be maintaining a map of nodes to replicas, or is that just more expensive?

**Comment by mrtracy** [comment]
I think the "mark as dead" is a natural dividing point for your service.  The timer should be separate from the operation.

Think of it this way: eventually, you're going to want better detection of down nodes than a five minute timer; there will be multiple signals that can declare a node to be dead.

I think you want a list of dead nodes with a "deadAt" timestamp.  The node comes back to life if that nodes "lastUpdateAt" is more recent than the "deadAt".

I also think you can just check every node on a regular interval; there's no reason to do advanced timer tricks, checking every 30 seconds won't be a burden.  You could even create a "node scanner" which is analogous to the replica scanner in each store.

File: docs/RFCS/range_repair.md (hunk 6 above)
Comment (BramGruneir) [inline docs/RFCS/range_repair.md]: interesting... might be worth considering as an optimization.

File: docs/RFCS/range_repair.md (hunk 5 above)
Comment (BramGruneir) [inline docs/RFCS/range_repair.md]: The last update time will be the time of `now` when the gossip callback trigger fires.

File: docs/RFCS/range_repair.md (hunk 4 above)
Comment (BramGruneir) [inline docs/RFCS/range_repair.md]: In the map, I was going to add a dead/alive bool, but you're right, this might not be needed.

File: docs/RFCS/range_repair.md (hunk 3 above)
Comment (BramGruneir) [inline docs/RFCS/range_repair.md]: I'll add it as a finally clause at the end.

**Comment by BramGruneir** [comment]
After discussions with Peter, I'm going to convert this to a `node health` or `node pool` that's part of gossip.  It will run the same loop and be triggered on node descriptor updates as the design up mentions, but it won't handle the replica removal.  Instead, it will maintain the mapping of node ids to health so the replica scanner can detect when a replica needs to be removed.

I am still a fan of the advanced timer, but I don't even know if it will be necessary. Going to look deep into gossip code first.
